src/ResourceManagement/ResourceManager/Authentication/MSITokenProvider.cs

```diff
@@ -65,47 +115,176 @@ public MSITokenProvider(string resource, MSILoginInformation msiLoginInformation
+                //
+                if (cache.TryGetValue(resource, out token) == true && !token.IsExpired)
+                {
+                    return token;
+                }
+                else
+                {
+                    token = await this.RetrieveTokenFromIMDSWithRetryAsync(resource, cancellationToken);
+                    cache.AddOrUpdate(resource, token, (key, oldValue) => token);
+                    return token;
+                }
             }
-            string accessToken = loginInfo.access_token;
-            if (accessToken == null)
+            finally
             {
-                throw MSILoginException.AcessTokenNotFound(content);
+                semaphoreSlim.Release();
+            }
+        }
+
+        private async Task<MSIToken> RetrieveTokenFromIMDSWithRetryAsync(string resource, CancellationToken cancellationToken)
+        {
+            var uriBuilder = new UriBuilder("http://169.254.169.254/metadata/identity/oauth2/token");
+            //
+            var query = new Dictionary<string, string>
+            {
+                ["api-version"] = "2018-02-01",
+                ["resource"] = resource
+            };
+            if (this.msiLoginInformation.UserAssignedIdentityObjectId != null)
+            {
+                query["object_id"] = this.msiLoginInformation.UserAssignedIdentityObjectId;
+            }
+            else if (this.msiLoginInformation.UserAssignedIdentityClientId != null)
+            {
+                query["client_id"] = this.msiLoginInformation.UserAssignedIdentityClientId;
+            }
+            else if (this.msiLoginInformation.UserAssignedIdentityResourceId != null)
+            {
+                query["msi_res_id"] = this.msiLoginInformation.UserAssignedIdentityResourceId;
+            }
+            else
+            {
+                throw new ArgumentException("MSI: UserAssignedIdentityObjectId, UserAssignedIdentityClientId or UserAssignedIdentityResourceId must be set");
+            }
+            uriBuilder.Query = await new FormUrlEncodedContent(query).ReadAsStringAsync();
+            string url = uriBuilder.ToString();
+            //
+            int retry = 1;
+            while (retry <= maxRetry)
+            {
+                //
+                using (HttpRequestMessage msiRequest = new HttpRequestMessage(HttpMethod.Get, url))
+                {
+                    msiRequest.Headers.Add("Metadata", "true");
+                    using (HttpResponseMessage msiResponse = await (new HttpClient()).SendAsync(msiRequest, cancellationToken))
+                    {
+                        int statusCode = ((int)msiResponse.StatusCode);
+                        if (ShouldRetry(statusCode))
+                        {
+
+                            int retryTimeout = retrySlots[new Random().Next(retry)];
+                            await Task.Delay(retryTimeout * 1000);
```

Comment:
>Task.Delay [](start = 34, length = 10)

no delays please :) use our extension method to remove delays in test mode (not sure if it is applicable here but we should have a consistent approach)